kotlinx-coroutines-core/common/src/CoroutineDispatcher.kt

```diff
@@ -27,46 +27,30 @@ import kotlin.coroutines.*
+public abstract class CoroutineDispatcher
+    : AbstractCoroutineContextElement(ContinuationInterceptor), ContinuationInterceptor {
+
     /**
-     * Returns `true` if the execution shall be dispatched onto another thread.
+     * Returns `true` if the execution of the coroutine should be performed with [dispatch] method.
      * The default behavior for most dispatchers is to return `true`.
      *
-     * This method should never be used from general code, it is used only by `kotlinx.coroutines`
-     * internals and its contract with the rest of the API is an implementation detail.
-     *
-     * UI dispatchers _should not_ override `isDispatchNeeded`, but leave the default implementation that
-     * returns `true`. To understand the rationale beyond this recommendation, consider the following code:
-     *
-     * ```kotlin
-     * fun asyncUpdateUI() = async(Dispatchers.Main) {
-     *     // do something here that updates something in UI
-     * }
-     * ```
-     *
-     * When you invoke `asyncUpdateUI` in some background thread, it immediately continues to the next
-     * line, while the UI update happens asynchronously in the UI thread. However, if you invoke
-     * it in the UI thread itself, it will update the UI _synchronously_ if your `isDispatchNeeded` is
-     * overridden with a thread check. Checking if we are already in the UI thread seems more
-     * efficient (and it might indeed save a few CPU cycles), but this subtle and context-sensitive
-     * difference in behavior makes the resulting async code harder to debug.
+     * If this method returns `false`, the coroutine is resumed immediately in the current thread,
+     * potentially forming an event-loop to prevent stack overflows.
+     * The event loop is an advanced topic and its implications can be found in [Dispatchers.Unconfined] documentation.
      *
-     * Basically, the choice here is between the "JS-style" asynchronous approach (async actions
-     * are always postponed to be executed later in the event dispatch thread) and "C#-style" approach
-     * (async actions are executed in the invoker thread until the first suspension point).
-     * While the C# approach seems to be more efficient, it ends up with recommendations like
-     * "use `yield` if you need to ....". This is error-prone. The JS-style approach is more consistent
-     * and does not require programmers to think about whether they need to yield or not.
+     * A dispatcher can override this method to provide a performance optimization and avoid paying a cost of an unnecessary dispatch.
+     * E.g. [MainCoroutineDispatcher.immediate] checks whether we are already in the required UI thread in this method and avoids
+     * an additional dispatch when it is not required. But this method should not return `false` by default, because
+     * it may expose unexpected behaviour (e.g. with interleaved event-loops) and unexpected order of events.
```

Comment:
This documentation now lacks rationale on why the default (non-immeidate) behavior works the way it is. There is no other place where it is currently documented. I'd suggest to either leave it here or move it somewhere.